lib/Sema/SemaBounds.cpp

```diff
@@ -628,93 +698,210 @@ namespace {
-            // indicate cause of failure to have bounds.
+            if (!LHSBounds->isNone() && !RHSBounds->isNone()) {
+              // TODO: Check if LHSBounds and RHSBounds are equal.
+              // if so, return one of them. If not, return bounds(none)
+              return CreateBoundsUnknown();
+            }
+            if (LHSBounds->isNone() && RHSBounds->isNone())
+              return CreateBoundsEmpty();
           }
-          return CreateBoundsNone();
+
+          // Comparisons and Logical Ops
+          //
+          // `e1 @ e2` have empty bounds if @ is:
+          // ==, !=, <=, <, >=, >, &&, ||
+          if (BinaryOperator::isComparisonOp(Op) ||
+              BinaryOperator::isLogicalOp(Op)) {
+            return CreateBoundsEmpty();
+          }
+
+          // All Other Binary Operators we don't know how to deal with
+          return CreateBoundsEmpty();
+        }
+        case Expr::CallExprClass: {
+          const CallExpr *CE = dyn_cast<CallExpr>(E);
+          if (!CE) {
+            llvm_unreachable("unexpected cast failure");
+            return CreateBoundsInferenceError();
+          }
+
+          // Get function prototype.
+          // The Callee is a function *pointer*, almost always.
+          const FunctionProtoType* CalleeTy = dyn_cast<FunctionProtoType>(CE->getCallee()->getType()->getPointeeType());
+          if (!CalleeTy)
+            // K&R functions have no prototype, and we cannot perform inference on them,
+            // so we return bounds(none) for their results.
+            return CreateBoundsEmpty();
+
+          BoundsExpr *FunBounds = const_cast<BoundsExpr *>(CalleeTy->getReturnBounds());
+          if (!FunBounds)
+            // This function has no return bounds
+            return CreateBoundsEmpty();
+
+          if (FunBounds->isInteropTypeAnnotation())
+            return CreateBoundsAllowedButUncomputable();
+
+          ArrayRef<Expr *> ArgExprs = llvm::makeArrayRef(const_cast<Expr**>(CE->getArgs()),
+                                                         CE->getNumArgs());
+
+          // Concretize Call Bounds with argument expressions.
+          // We can only do this if the argument expressions are non-modifying
+          BoundsExpr *ReturnBounds = SemaRef.ConcretizeFromFunctionTypeWithArgs(FunBounds, ArgExprs);
+
+          // If concretization failed, we can't yet compute static bounds
+          // for the return type of this function.
+          if (!ReturnBounds)
+            return CreateBoundsAllowedButUncomputable();
+
+          // Currently we cannot yet concretize function bounds of the forms
+          // count(e) or byte_count(e) becuase we need a way of referring
+          // to the function's return value which we currently lack in the
+          // general case.
+          if (ReturnBounds->isElementCount() ||
+              ReturnBounds->isByteCount())
+            return CreateBoundsAllowedButUncomputable();
```

Comment:
We also can't compute the right bounds at the moment if they're `count(n)` or `byte_count(n)` without being able to expand these to use `return_result` or similar.